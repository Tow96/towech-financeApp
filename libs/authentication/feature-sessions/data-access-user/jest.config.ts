/* eslint-disable */
export default {
  displayName: 'authentication-feature-sessions-data-access-user',
  preset: '../../../../jest.preset.js',
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../../../coverage/libs/authentication/feature-sessions/data-access-user',
  coverageReporters: ['html', 'text', 'text-summary', 'cobertura', 'lcov', 'json'],
  collectCoverageFrom: ['**/*.ts', '!**/*.module.ts', '!**/index.ts', '!**/jest.config.ts'],
};
