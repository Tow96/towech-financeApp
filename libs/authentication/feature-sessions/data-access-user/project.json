{
  "name": "authentication-feature-sessions-data-access-user",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/authentication/feature-sessions/data-access-user/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/authentication/feature-sessions/data-access-user/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/authentication/feature-sessions/data-access-user/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
