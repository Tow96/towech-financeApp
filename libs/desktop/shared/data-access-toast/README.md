# desktop-shared-data-access-toast

This library was generated with [Nx](https://nx.dev).


## Running unit tests

Run `nx test desktop-shared-data-access-toast` to execute the unit tests.

