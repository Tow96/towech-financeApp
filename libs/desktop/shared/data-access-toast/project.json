{
  "name": "desktop-shared-data-access-toast",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/desktop/shared/data-access-toast/src",
  "prefix": "finance",
  "tags": [],
  "projectType": "library",
  "targets": {
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/desktop/shared/data-access-toast/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": [
          "libs/desktop/shared/data-access-toast/**/*.ts",
          "libs/desktop/shared/data-access-toast/**/*.html"
        ]
      }
    }
  }
}
