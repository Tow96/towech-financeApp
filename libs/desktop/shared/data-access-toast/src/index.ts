export * from './lib/toast.service';
