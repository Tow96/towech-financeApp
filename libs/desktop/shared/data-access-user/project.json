{
  "name": "desktop-shared-data-access-user",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/desktop/shared/data-access-user/src",
  "prefix": "finance",
  "tags": [],
  "projectType": "library",
  "targets": {
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/desktop/shared/data-access-user/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": [
          "libs/desktop/shared/data-access-user/**/*.ts",
          "libs/desktop/shared/data-access-user/**/*.html"
        ]
      }
    }
  }
}
