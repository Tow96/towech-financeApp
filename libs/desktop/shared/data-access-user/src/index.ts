export * from './lib/user.service';
export * from './lib/types';
