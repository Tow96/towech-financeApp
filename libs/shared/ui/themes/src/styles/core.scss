// The theming functions are heavily taken from angular material, but in a lighter way
// Original code: https://material.angular.io/
@use 'sass:map';

@function get-palette-hue($palette, $hue) {
  @return map.get($palette, $hue);
}

@function get-palette-hue-contrast($palette, $hue) {
  @return map.get(map.get($palette, contrast), $hue);
}

@function define-theme(
  $background,
  $text,
  $primary,
  $accent,
  $success: null,
  $warning: null,
  $error: null
) {
  @if $background == null or $text == null {
    @error 'Background and text colors need to be defined';
  }

  @if $primary == null or $accent == null {
    @error 'primary and accent colors need to be defined palettes';
  }

  @if not map.get($primary, contrast) {
    @error 'Primary is not a palette';
  }

  @if not map.get($accent, contrast) {
    @error 'Accent is not a palette';
  }

  @if $success != null and not map.get($success, contrast) {
    @error 'Success is not a palette';
  }

  @if $error != null and not map.get($error, contrast) {
    @error 'Error is not a palette';
  }

  @if $warning != null and not map.get($warning, contrast) {
    @error 'Warning is not a palette';
  }

  $result: (
    color: (
      text: $text,
      background: $background,
      primary: $primary,
      accent: $accent,
      success: if($success != null, $success, $accent),
      warning: if($warning != null, $warning, $accent),
      error: if($error != null, $error, $accent),
    ),
  );

  @return $result;
}

@mixin color($theme) {
  $colors: map.get($theme, color);
  $primary: map.get($colors, primary);

  :root {
    background-color: map.get($colors, background);
    color: map.get($colors, text);
  }

  div.container {
    background-color: get-palette-hue($primary, 500);
  }
}
