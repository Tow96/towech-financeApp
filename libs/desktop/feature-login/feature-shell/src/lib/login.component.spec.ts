// Libraries
import { Component, EventEmitter } from '@angular/core';
import { ComponentFixture, TestBed, fakeAsync, tick } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
import { AsyncPipe } from '@angular/common';
import { SubscriberSpy, subscribeSpyTo } from '@hirez_io/observer-spy';
// Services
import { DesktopUserService } from '@finance/desktop/shared/data-access-user';
import { DesktopToasterService } from '@finance/desktop/shared/data-access-toast';
// Tested elements
import { DesktopLoginComponent, PatchFormGroupValuesDirective } from './login.component';
// Mocks
import {
  fakeUserServiceFire,
  DesktopToasterServiceMock,
  DesktopUserServiceMock,
} from '@finance/desktop/shared/utils-testing';
import { provideStore } from '@ngrx/store';
import { Action, adaptReducer } from '@state-adapt/core';

// Directive ------------------------------------------------------------------
@Component({
  standalone: true,
  imports: [PatchFormGroupValuesDirective, ReactiveFormsModule, AsyncPipe],
  template: ` <form [formGroup]="form" [patchFormGroupValues]="patcher$ | async"></form> `,
})
class TestComponent {
  public form = new FormGroup({
    user: new FormControl(''),
    pass: new FormControl(''),
  });

  public patch = new EventEmitter<{ user: string; pass: string }>();
  public patcher$ = this.patch.pipe();
}

describe('Patch form Directive', () => {
  let component: TestComponent;
  let fixture: ComponentFixture<TestComponent>;
  let spy: jest.SpyInstance;

  beforeEach(() => {
    jest.clearAllMocks();
    TestBed.configureTestingModule({ imports: [TestComponent] });
    fixture = TestBed.createComponent(TestComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    spy = jest.spyOn(component.form, 'patchValue');
  });

  describe('When the patch pipe is called without anything', () => {
    it('Should not modify the form', () => {
      const prevValue = { ...component.form.value };
      // component.patcher$.next(undefined);
      const newValue = { ...component.form.value };

      expect(prevValue).toEqual(newValue);
      expect(spy).toHaveBeenCalledTimes(0);
    });
  });

  describe('When the patch pipe is called with an update', () => {
    it('Should patch the form and have it NOT emit', () => {
      const newValue = { user: 'newuser', pass: 'password' };
      component.patch.next(newValue);
      fixture.detectChanges();
      expect(component.form.value).toEqual(newValue);
      expect(spy).toHaveBeenCalledWith(newValue, { emitEvent: false });
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });
});

// Component ------------------------------------------------------------------
describe('Desktop Login Component', () => {
  let component: DesktopLoginComponent;
  let fixture: ComponentFixture<DesktopLoginComponent>;
  let userService: DesktopUserService;
  let toastService: DesktopToasterService;
  let compiled: HTMLElement;
  let stateSpy: SubscriberSpy<any>;

  const updateComponent = () => {
    fixture.detectChanges();
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
  };

  beforeEach(() => {
    jest.clearAllMocks();
    TestBed.configureTestingModule({
      providers: [
        provideStore({ adapt: adaptReducer }),
        { provide: DesktopToasterService, useValue: DesktopToasterServiceMock },
        { provide: DesktopUserService, useValue: DesktopUserServiceMock },
      ],
    });
    fixture = TestBed.createComponent(DesktopLoginComponent);

    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;

    userService = TestBed.inject<DesktopUserService>(DesktopUserService);
    toastService = TestBed.inject<DesktopToasterService>(DesktopToasterService);
    stateSpy = subscribeSpyTo(component.store.state$);
  });

  it('Should be defined', () => expect(component).toBeTruthy());

  it('Must match the snapshot', () => expect(compiled).toMatchSnapshot());

  describe('When the form fields are changed', () => {
    it('Should update the store after the debounce time has passed', fakeAsync(() => {
      const newState = {
        ...component.initialState,
        keepSession: false,
        password: 'pass',
        username: 'user',
      };

      component.form.patchValue(newState);
      tick(1000);
      expect(stateSpy.getLastValue()).toEqual(newState);
    }));
  });

  describe('When the login button is pressed', () => {
    let userSpy: SubscriberSpy<Action<any, string>>;
    let toastSpy: SubscriberSpy<Action<any, string>>;
    let loginBttn: HTMLElement;

    beforeEach(() => {
      userSpy = subscribeSpyTo(userService.login$);
      toastSpy = subscribeSpyTo(toastService.addError$);
      loginBttn = fixture.debugElement.nativeElement.querySelector('#Login-button > button');
    });

    describe('With an invalid form', () => {
      beforeEach(() => {
        component.form.patchValue({ username: '', keepSession: false, password: '' });
        loginBttn.click();
      });

      it('Should not call the user service', () => expect(userSpy.receivedNext()).toBeFalsy());

      it('Should send an error toast', () => expect(toastSpy.receivedNext()).toBeTruthy());
    });

    describe('With a valid form', () => {
      it('Should next the userservice', () => {
        component.form.patchValue({ username: 'user', keepSession: false, password: 'pass' });
        loginBttn.click();
        expect(userSpy.receivedNext()).toBe(true);
      });
    });
  });

  describe('When the userService nexts the onLoginError$ pipe', () => {
    beforeEach(() =>
      fakeUserServiceFire.next({ payload: '', type: '[User Service] Login.error$' })
    );

    it('Should clear the form', () => expect(stateSpy.getLastValue()).toBe(component.initialState));

    it('Should set the form components to dirty', () => {
      updateComponent();
      expect(fixture.debugElement.query(By.css('#form-username')).classes['ng-dirty']).toBeTruthy();
      expect(fixture.debugElement.query(By.css('#form-password')).classes['ng-dirty']).toBeTruthy();
      expect(fixture.debugElement.query(By.css('#form-keep')).classes['ng-dirty']).toBeTruthy();
    });
  });
});
