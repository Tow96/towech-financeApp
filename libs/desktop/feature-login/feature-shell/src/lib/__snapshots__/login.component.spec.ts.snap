// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desktop Login Component Must match the snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="login"
  >
    
    <div
      class="login__container"
    >
      <h1>
        Login
      </h1>
      <finance-login-form>
        <form
          class="ng-untouched ng-pristine ng-invalid"
          novalidate=""
        >
          <finance-input
            class="ng-untouched ng-pristine ng-invalid"
            formcontrolname="username"
            id="form-username"
            label="Username"
          >
            <div
              class="input_component"
            >
              <input
                placeholder=" "
                type="text"
              />
              <label>
                Username
              </label>
            </div>
          </finance-input>
          <finance-input
            class="ng-untouched ng-pristine ng-invalid"
            formcontrolname="password"
            id="form-password"
            label="Password"
            type="password"
          >
            <div
              class="input_component"
            >
              <input
                placeholder=" "
                type="password"
              />
              <label>
                Password
              </label>
            </div>
          </finance-input>
          <div
            class="bottom-row"
          >
            <finance-checkbox
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="keepSession"
              id="form-keep"
              label="Keep Session"
            >
              <div
                class="checkbox"
              >
                <input
                  id="Keep Session"
                  type="checkbox"
                />
                <label
                  for="Keep Session"
                >
                  Keep Session
                </label>
              </div>
            </finance-checkbox>
            <finance-button
              id="Login-button"
              type="submit"
            >
              <button
                class="accent"
                type="submit"
              >
                Login
              </button>
            </finance-button>
          </div>
        </form>
        
      </finance-login-form>
    </div>
  </div>
</div>
`;
